exchanges/kucoin/kucoin_websocket.go

```diff
@@ -0,0 +1,1069 @@
+	if err != nil {
+		return err
+	}
+	ku.Websocket.DataHandler <- &order.Detail{
+		Price:           response.Price,
+		Amount:          response.Size,
+		ExecutedAmount:  response.FilledSize,
+		RemainingAmount: response.RemainSize,
+		Exchange:        ku.Name,
+		OrderID:         response.OrderID,
+		ClientOrderID:   response.ClientOid,
+		Type:            oType,
+		Side:            side,
+		Status:          oStatus,
+		AssetType:       asset.Spot,
+		Date:            response.OrderTime.Time(),
+		LastUpdated:     response.Timestamp.Time(),
+		Pair:            pair,
+	}
+	return nil
+}
+
+func (ku *Kucoin) processTradeData(respData []byte, instrument string) error {
+	response := WsTrade{}
+	err := json.Unmarshal(respData, &response)
+	if err != nil {
+		return err
+	}
+	saveTradeData := ku.IsSaveTradeDataEnabled()
+	if !saveTradeData &&
+		!ku.IsTradeFeedEnabled() {
+		return nil
+	}
+	pair, err := currency.NewPairFromString(instrument)
+	if err != nil {
+		return err
+	}
+	side, err := order.StringToOrderSide(response.Side)
+	if err != nil {
+		return err
+	}
+	return ku.Websocket.Trade.Update(saveTradeData, trade.Data{
+		CurrencyPair: pair,
+		Timestamp:    time.UnixMilli(response.Time),
+		Price:        response.Price,
+		Amount:       response.Size,
+		Side:         side,
+		Exchange:     ku.Name,
+		TID:          response.TradeID,
+		AssetType:    asset.Spot,
+	})
+}
+
+func (ku *Kucoin) processTicker(respData []byte, instrument string) error {
+	response := WsTicker{}
+	err := json.Unmarshal(respData, &response)
+	if err != nil {
+		return err
+	}
+	pair, err := currency.NewPairFromString(instrument)
+	if err != nil {
+		return err
+	}
+	ku.Websocket.DataHandler <- &ticker.Price{
+		AssetType:    asset.Spot,
+		Last:         response.Size,
+		LastUpdated:  time.Now(),
+		ExchangeName: ku.Name,
+		Pair:         pair,
+		Ask:          response.BestAsk,
+		Bid:          response.BestBid,
+		AskSize:      response.BestAskSize,
+		BidSize:      response.BestBidSize,
+	}
+	return nil
+}
+
+func (ku *Kucoin) processCandlesticks(respData []byte, instrument, intervalString string) error {
+	pair, err := currency.NewPairFromString(instrument)
+	if err != nil {
+		return err
+	}
+	response := WsCandlestickData{}
+	err = json.Unmarshal(respData, &response)
+	if err != nil {
+		return err
+	}
+	resp, err := response.getCandlestickData()
+	if err != nil {
+		return err
+	}
+	ku.Websocket.DataHandler <- stream.KlineData{
+		Timestamp:  time.UnixMilli(response.Time),
+		Pair:       pair,
+		AssetType:  asset.Spot,
+		Exchange:   ku.Name,
+		StartTime:  resp.Candles.StartTime,
+		Interval:   intervalString,
+		OpenPrice:  resp.Candles.OpenPrice,
+		ClosePrice: resp.Candles.ClosePrice,
+		HighPrice:  resp.Candles.HighPrice,
+		LowPrice:   resp.Candles.LowPrice,
+		Volume:     resp.Candles.TransactionVolume,
+	}
+	return nil
+}
+
+func (ku *Kucoin) processOrderbook(respData []byte, subject, instrument string) error {
+	response := WsOrderbook{}
```

Comment:
This does not look to be implementing the guidelines here: https://docs.kucoin.com/#level-2-market-data